src/core/common/timer.hpp

```diff
@@ -128,16 +144,60 @@ class TimerScheduler
      */
     static bool IsStrictlyBefore(uint32_t aTimeA, uint32_t aTimeB);
 
-    Timer *mHead;
+    TimerBase *mHead;
+};
+
+/**
+ * This class implements the millisecond timer scheduler.
+ *
+ */
+class TimerMilliScheduler: public TimerSchedulerBase
+{
+public:
+    /**
+     * This constructor initializes the object.
+     *
+     * @param[in]  aIp6  A reference to the IPv6 network object.
+     *
+     */
+    TimerMilliScheduler(Ip6::Ip6 &aIp6):
+        TimerSchedulerBase(aIp6) {
+    }
+
+    /**
+     * This method adds a timer instance to the timer scheduler.
+     *
+     * @param[in]  aTimer  A reference to the timer instance.
+     *
+     */
+    void Add(TimerMilli &aTimer) { TimerSchedulerBase::Add(reinterpret_cast<TimerBase &>(aTimer), sAlarmMilliApi); }
```

Comment:
I was thinking if we can avoid the `reinterpret_cast<>` here but may be tricky (as complier reaching here does not yet know `TimerMilli` inherits from `TimerBase`). We can consider changing the order of declaration of `TimerMilliShceduler` and `TimerMilli` classes. It then creates problem that the we cannot use `TimerMilliShceduler` methods in inline definitions of `StartAt()` and `Stop()` but I think we can move the definitions into the `.cpp` file.